bin/packages/check-build-type-declaration-files.js

```diff
@@ -0,0 +1,114 @@
+/**
+ * The purpose of this script is to double-check certain published build types.
+ * When we translate JSDoc into types that we wish to publish, we occasionally
+ * disable the type checker for JS files. Unfortunately, this can result in invalid
+ * index.d.ts files being published.
+ *
+ * This script verifies the publishd index.d.ts file for every package which both
+ * builds types and also sets checkJs to false in its tsconfig.json. This is done
+ * by running `tsc --noEmit` on the $package/build-types/index.d.ts file. This also
+ * verifies everything index.d.ts references, which essentially checks the entire
+ * public api of the type declarations for that package.
+ *
+ * @see https://github.com/WordPress/gutenberg/pull/49650 for more discussion.
+ */
+
+/**
+ * External dependencies
+ */
+const fs = require( 'fs' ).promises;
+const path = require( 'path' );
+const { exec } = require( 'child_process' );
+const chalk = require( 'chalk' );
+
+// Tsconfig can be JSON5, which accepts comments, which need to be stripped so we can parse the JSON.
+function stripJsonComments( jsonString ) {
+	const lines = jsonString.split( '\n' );
+	const cleanedLines = lines.map( ( line ) => line.replace( /\/\/.*$/, '' ) );
+	return cleanedLines.join( '\n' );
```

Comment:
it's probably fine here practically speaking, but this chokes on strings containing `//`, also it could be done in a single RegExp replace if we care. I didn't comment because the line orientation has its clarity too

```js
function stripJsonComments( jsonString ) {
	// Strip everything past the `//` to the end of a line.
	return string.replace( /\s+\/\/.*$/gm, '' );
}
```
https://regex101.com/r/J5CDCS/1

If we were to run into any problems we would probably want to reach for a JSON parser, which isn't that complicated to write, and which some may exist, but we might as well also use our own if we need.

Doubtful we'll need it.

Reply:
I tried getting GPT-3 to write a proper parser but it's been a long failing endeavor. it can't seem to get all the basic cases, and really struggles with comment syntax found inside strings, or escaped quotes inside strings.

Reply:
finally got there with a working parser. had to ask it to provide test inputs and verify them.

Reply:
I think that's a good point. Worth adding a TODO if we decide not to do it immediately.

Reply:
> If we were to run into any problems we would probably want to reach for a JSON parser, which isn't that complicated to write, and which some may exist, but we might as well also use our own if we need.

Its original suggestion was to introduce json5 (https://www.npmjs.com/package/json5), but I felt we just didn't need to be that heavy-handed here. 

I'll add a comment and use this better regex, though!

Reply:
> we just didn't need to be that heavy-handed here

agreed, in case my thoughts were unclear too. a simple comment noting the simplification should suffice.